crates/storage/src/ops/envelope.rs

```diff
@@ -0,0 +1,84 @@
+//! Operations for reading/writing envelope related data from/to Database
+
+use std::sync::Arc;
+
+use strata_db::{
+    traits::{L1PayloadDatabase, SequencerDatabase},
+    types::PayloadEntry,
+    DbResult,
+};
+use strata_primitives::buf::Buf32;
+use threadpool::ThreadPool;
+
+use crate::exec::*;
+
+/// Database context for an database operation interface.
+pub struct Context<D> {
+    db: Arc<D>,
+}
+
+impl<D: SequencerDatabase + Sync + Send + 'static> Context<D> {
+    /// Create a `Context` for [`EnvelopeDataOps`]
+    pub fn new(db: Arc<D>) -> Self {
+        Self { db }
+    }
+
+    /// Convert to [`EnvelopeDataOps`] using a [`ThreadPool`]
+    pub fn into_ops(self, pool: ThreadPool) -> EnvelopeDataOps {
+        EnvelopeDataOps::new(pool, Arc::new(self))
+    }
+}
+
+inst_ops! {
+    (EnvelopeDataOps, Context<D: SequencerDatabase>) {
+        get_payload_entry(id: Buf32) => Option<PayloadEntry>;
+        get_payload_entry_by_idx(idx: u64) => Option<PayloadEntry>;
+        get_payload_entry_id(idx: u64) => Option<Buf32>;
+        get_next_payload_idx() => u64;
+        put_payload_entry(id: Buf32, entry: PayloadEntry) => ();
+    }
+}
+
+fn get_payload_entry<D: SequencerDatabase>(
+    ctx: &Context<D>,
+    id: Buf32,
+) -> DbResult<Option<PayloadEntry>> {
+    let payload_db = ctx.db.payload_db();
+    payload_db.get_payload_by_id(id)
+}
+
+fn get_payload_entry_id<D: SequencerDatabase>(
+    ctx: &Context<D>,
+    idx: u64,
+) -> DbResult<Option<Buf32>> {
+    let payload_db = ctx.db.payload_db();
+    payload_db.get_payload_id(idx)
+}
+
+fn get_payload_entry_by_idx<D: SequencerDatabase>(
+    ctx: &Context<D>,
+    idx: u64,
+) -> DbResult<Option<PayloadEntry>> {
+    let payload_db = ctx.db.payload_db();
+    let id_res = payload_db.get_payload_id(idx)?;
+    match id_res {
+        Some(id) => payload_db.get_payload_by_id(id),
+        None => Ok(None),
+    }
+}
+
+fn get_next_payload_idx<D: SequencerDatabase>(ctx: &Context<D>) -> DbResult<u64> {
+    let payload_db = ctx.db.payload_db();
+    payload_db
+        .get_last_payload_idx()
+        .map(|x| x.map(|i| i + 1).unwrap_or_default())
+}
```

Comment:
This is kinda jank, we should do something to make this more consistent/safe.

Reply:
Yes. Looks ugly. Changed to 
```rust
Ok(payload_db
        .get_last_payload_idx()?
        .map(|i| i + 1)
        .unwrap_or_default())
```

Reply:
I was referring to how we're juggling the concepts of "next" and "last" here between the layers and doing the +1/-1 math here instead of being consistent throughout.

Reply:
It does make sense for the underlying db to consistently provide the "last" idx and the layer downstream can use it as it needs. Some layers need the last and some need the next, I can't see a problem here.

Reply:
The index of the "last element" of a growing sequence isn't well-defined if there's no elements, but "next" is, it's 0.  With "last" we have to special-case where there's no elements, and return `None` or something, meaning the interface can't simply return an index it has to return an `Option<_>` around the index type.

The lowest-level interface for a store like this should usually be providing the "next" index.

Reply:
That calls for another ticket, there are a lot of places where we've exposed the last inserted id from database.